Rename Indexer client to Generator client throughout documentation (#253)
shinzonetwork/docs: content/guides/operator-quickstart/index.md

+++
title = "Operator Quickstart"

> description = "Bring up an Generator and a Host on one machine, peer them over libp2p, and query the result."

Reviewer: ```suggestion
description = "Bring up a Generator client and a Host on one machine, peer them over libp2p, and query the result."
```

weight = 3
[extra]
mermaid = true
+++

> Run a Shinzo Generator and a Host on the same machine, peer them over libp2p, and query indexed Ethereum data through the Host's GraphQL API. If your Geth node is already reachable, the whole thing takes about ten minutes.

Reviewer: ```suggestion
Run a Shinzo Generator client and a Host client on the same machine, peer them over libp2p, and query indexed Ethereum data through the Host's GraphQL API. If your Geth node is already reachable, the whole thing takes about ten minutes.
```

When you're done you'll have:

> - A running Generator pulling blocks from a Geth node and signing them.

Reviewer: ```suggestion
- A running Generator client pulling blocks from a Geth node and signing them.
```

{% mermaid() %}
flowchart LR
  Geth["Geth (your node)"] -->|RPC + WS| Generator

>   Generator -->|libp2p| Host

Reviewer: ```suggestion
  Geth["Geth (your node)"] -->|RPC + WS| Generator client
  Generator client -->|libp2p| Host
```

  Host -->|GraphQL| You["You (curl)"]
{% end %}

> Two containers through one shared Docker bridge. Both containers run on the same VM. The Host dials the Generator's libp2p port directly over the bridge using the Generator's published Peer ID.

Reviewer: ```suggestion
Two containers through one shared Docker bridge. Both containers run on the same VM. The Host dials the Generator client's libp2p port directly over the bridge using the Generator client's published Peer ID.
```

## Prerequisites

- Docker. 
- Both `curl` and `jq`.

> - A reachable Ethereum execution node (Geth or compatible) exposing JSON-RPC and WebSocket. The Generator reads from this node; it does not run one for you. Acceptable sources include a node you self-host, a node co-located with a validator, GCP Blockchain Node Engine, or a managed provider like Alchemy or QuickNode. If your node is behind authentication, see the [Generator install guide's notes on API keys](/generator/install#do-you-need-an-api-key).

Reviewer: ```suggestion
- A reachable Ethereum execution node (Geth or compatible) exposing JSON-RPC and WebSocket. The Generator client reads from this node; it does not run one for you. Acceptable sources include a node you self-host, a node co-located with a validator, GCP Blockchain Node Engine, or a managed provider like Alchemy or QuickNode. If your node is behind authentication, see the [Generator client install guide's notes on API keys](/generator/install#do-you-need-an-api-key).
```

```

{% admonition(type="tip") %}

> The Generator auto-detects the right header (`X-goog-api-key` for GCP Blockchain Node Engine, `X-Api-Key` for most self-hosted Nginx setups) based on the URL. If the node is on your private network and unauthenticated, leave `GETH_API_KEY` empty.

Reviewer: ```suggestion
The Generator client auto-detects the right header (`X-goog-api-key` for GCP Blockchain Node Engine, `X-Api-Key` for most self-hosted Nginx setups) based on the URL. If the node is on your private network and unauthenticated, leave `GETH_API_KEY` empty.
```

{% end %}

> ## Start the Generator

Reviewer: ```suggestion
## Start the Generator client
```

> The Generator sits next to a blockchain node, subscribes to new blocks, and turns them into signed structured documents. Run it as a single container with three ports exposed:

Reviewer: ```suggestion
The Generator client sits next to a blockchain node, subscribes to new blocks, and turns them into signed structured documents. Run it as a single container with three ports exposed:
```

| Host port | Container port | What it is |
| --- | --- | --- |
| `9181` | `9181` | DefraDB GraphQL API. The query interface for raw indexed data. |

> | `9171` | `9171` | libp2p P2P port. How Hosts subscribe to this Generator. |

Reviewer: ```suggestion
| `9171` | `9171` | libp2p P2P port. This is how Hosts subscribe to the Generator client. |
```

  -p 9181:9181 \
  -p 9171:9171 \
  -p 8080:8080 \
  ghcr.io/shinzonetwork/shinzo-generator-client:standard
```

> `DEFRADB_KEYRING_SECRET` is the password that protects the Generator's signing key. The Generator uses this key to sign every document it produces, which gives downstream consumers a way to verify the data came from a real Generator. The `devnet-secret` password is fine for this quickstart, but remember use something more secure for anything in a production environment.

Reviewer: ```suggestion
`DEFRADB_KEYRING_SECRET` is the password that protects the Generator client's signing key. The Generator client uses this key to sign every document it produces, which gives downstream consumers a way to verify the data came from a real Generator client. The `devnet-secret` password is fine for this quickstart, but remember use something more secure for anything in a production environment.
```

`DEFRADB_P2P_LISTEN_ADDR` tells DefraDB which interface and port to bind libp2p to inside the container. Binding to `0.0.0.0:9171` means the Host container, running on the same Docker bridge, can reach it.

`INDEXER_START_HEIGHT=0` starts indexing from genesis. For Ethereum Mainnet that's a lot of history, but this quickstart only needs a few committed blocks to verify the pipeline. To start closer to head, set this to a recent block height.

`DEFRADB_PLAYGROUND=true` enables a browser-based GraphQL playground on the API port.

> ## Read the Generator's P2P address

Reviewer: ```suggestion
## Read the Generator client's P2P address
```

The Host needs two things to connect: 

> 1. The Generator's libp2p Peer ID.

Reviewer: ```suggestion
1. The Generator client's libp2p Peer ID.
```

1. A multiaddr it can dial.

> Both come from the Generator's `/health` endpoint once it finishes starting up.

Reviewer: ```suggestion
Both come from the Generator client's `/health` endpoint once it finishes starting up.
```

```
{% end %}

> `id` is the Generator's libp2p Peer ID, derived from its keyring secret. It's stable across restarts as long as the secret stays the same.

Reviewer: ```suggestion
`id` is the Generator client's libp2p Peer ID, derived from its keyring secret. It's stable across restarts as long as the secret stays the same.
```

> `addresses` are the multiaddrs the Generator is actually listening on. The first one (`127.0.0.1`) is loopback and useless to other containers. The second (`172.17.0.2` here) is the Generator container's IP on the Docker bridge. That's the one the Host will dial.

Reviewer: ```suggestion
`addresses` are the multiaddrs the Generator client is actually listening on. The first one (`127.0.0.1`) is loopback and useless to other containers. The second (`172.17.0.2` here) is the Generator client container's IP on the Docker bridge. That's the one the Host will dial.
```

There are a few things in here worth calling out:

> - `defradb.url: localhost:9181` is the Host's _internal_ DefraDB API, not the Generator's. Inside the Host container, DefraDB binds to `9181` on `localhost`. The Generator's API happens to use the same number because they're both DefraDB; we'll remap the published ports in the next step so they don't collide.

Reviewer: ```suggestion
- `defradb.url: localhost:9181` is the Host's _internal_ DefraDB API, not the Generator client's. Inside the Host container, DefraDB binds to `9181` on `localhost`. The Generator client's API happens to use the same number because they're both DefraDB; we'll remap the published ports in the next step so they don't collide.
```

> - `bootstrap_peers` is the only Generator-specific value. The Host learns everything else (schemas, signed data) from the Generator over P2P once it connects.

Reviewer: ```suggestion
- `bootstrap_peers` is the only Generator-specific value. The Host learns everything else (schemas, signed data) from the Generator client over P2P once it connects.
```

> - `minimum_attestations: 1` means the Host will serve data as soon as it has one signature on it. Production setups use higher values to require independent confirmation from multiple Generators. See the [Host overview](/hosts/overview) for more on attestations.

Reviewer: ```suggestion
- `minimum_attestations: 1` means the Host will serve data as soon as it has one signature on it. Production setups use higher values to require independent confirmation from multiple Generator clients. See the [Host overview](/hosts/overview) for more on attestations.
```

- `hub_base_url` points to ShinzoHub. We're not registering anything here, but the Host expects the field to be present.

## Start the Host

> The Generator is already occupying `9181`, `9171`, and `8080` on the host machine, so the Host container's ports get bumped by one.

Reviewer: ```suggestion
The Generator client is already occupying `9181`, `9171`, and `8080` on the host machine, so the Host container's ports get bumped by one.
```

## Verify the peering

> The Host's `/health` lists the peers it's connected to. Once the Generator's Peer ID appears there, the connection is live.

Reviewer: ```suggestion
The Host's `/health` lists the peers it's connected to. Once the Generator client's Peer ID appears there, the connection is live.
```

```
{% end %}

> Check the Generator's side too. Its peer list should now contain the Host:

Reviewer: ```suggestion
Check the Generator client's side too. Its peer list should now contain the Host:
```

```
{% end %}

> If both list each other, libp2p is connected and DefraDB is replicating between them! Data from the Generator lands in the Host within a few seconds.

Reviewer: ```suggestion
If both list each other, libp2p is connected and DefraDB is replicating between them! Data from the Generator client lands in the Host within a few seconds.
```

```
{% end %}

> The rows that come back were originally just logs on Ethereum, then pulled in by the Generator over the Geth WebSocket, signed, gossiped over libp2p to the Host, and are now being served back to you over GraphQL. More queries are on the [Host examples](/hosts/examples) page.

Reviewer: ```suggestion
The rows that come back were originally just logs on Ethereum, then pulled in by the Generator client over the Geth WebSocket, signed, gossiped over libp2p to the Host, and are now being served back to you over GraphQL. More queries are on the [Host examples](/hosts/examples) page.
```

## Undo everything

If you want to burn everything down and start again, just run:

```shell
docker rm -f shinzo-generator shinzo-host
rm ~/host-config.yaml
```

## Troubleshooting

### The Generator's `/health` never returns a Peer ID

```shell
docker logs --tail 50 shinzo-generator
```

> The most common cause is the Generator can't reach Geth. Confirm `GETH_RPC_URL` and `GETH_WS_URL` are correct and reachable from the container, and that the API key (if any) is right. The Generator can fall back to HTTP-only mode if WebSocket is unavailable, and will log that it did.

Reviewer: ```suggestion
The most common cause is the Generator client can't reach Geth. Confirm `GETH_RPC_URL` and `GETH_WS_URL` are correct and reachable from the container, and that the API key (if any) is right. The Generator client can fall back to HTTP-only mode if WebSocket is unavailable, and will log that it did.
```

### The Host never lists the Generator as a peer

> Confirm the bootstrap peer multiaddr in the config uses the Generator's non-loopback container IP, not `127.0.0.1`. From inside the Host container, the Generator's loopback is unreachable. Verify with:

Reviewer: ```suggestion
Confirm the bootstrap peer multiaddr in the config uses the Generator client's non-loopback container IP, not `127.0.0.1`. From inside the Host container, the Generator client's loopback is unreachable. Verify with:
```

### Port already in use

> The Generator or the Host is colliding with something else on the machine. Change the published ports in the `docker run` commands. Container-side ports stay the same.

Reviewer: ```suggestion
The Generator client or Host client is colliding with something else on the machine. Change the published ports in the `docker run` commands. Container-side ports stay the same.
```
